### DISCUSSION TITLE
Access appsettings.json in external module and register in services

### DISCUSSION THREAD

**Discussion by qmmughal** [oqtane/oqtane.framework#1384] [Q&A]
I am looking to read appsettings and register in services as shown below. 
I got NULL reference error for Configuration when i try to run app from Oqtane Solution. any hint?

 public class ServerStartup : IServerStartup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            services.Configure<AppSettingsOptions>(Configuration.GetSection("AppSettings"))
                 .AddSingleton(x => x.GetRequiredService<IOptions<AppSettingsOptions>>().Value);
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            
        }

        public void ConfigureMvc(IMvcBuilder mvcBuilder)
        {
        }
    }

**Comment by sbwalker** [comment]
I assume you are asking about how to access the Configuration through the Server aspect of your module - as you cannot access Configuration from the Client side ( ie. razor components ) directly. Configuration is already registered by Oqtane in DI so you simply need to inject it into your Server class:

```
    public class YourClass
    {
        private readonly IConfigurationRoot _config;

        public YourClass(IConfigurationRoot config)
        {
            _config = config;
        }
```

And then you can use _config.GetSection(), etc...

**Comment by Gotiap** [comment]
In this case let's say I want to register http service in startup and I need to use different url from production, staging and dev environment. So I need to have access that into ConfigureServices method from StartupClass.

@sbwalker 
Can you please suggest on this?

Thank you

**Reply by sbwalker** [reply]
Create an Oqtane module. Include your services in the Client project. Decorate your services with the IService interface and they will be automaticallty registered during startup for you (look at HtmlTextService in the framework). Your services will not be able to access appsettings.json as they are running on the client and appsettings.json is on the server. So you will need to find another approach for dealing with urls - perhaps storing them in module settings would be  an option.

**Comment by ADefWebserver** [comment]
This worked for me:
https://github.com/ADefWebserver/Oqtane.SyncfusionHelpDesk/edit/main/Server/Startup/ServerStartup.cs

```
﻿using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.DependencyInjection;
using Oqtane.Infrastructure;
using Syncfusion.Blazor;
using Microsoft.Extensions.Configuration;
namespace Syncfusion.Helpdesk.Server.Startup
{
    public class ServerStartup : IServerStartup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddSyncfusionBlazor();
        }
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            // Register Syncfusion license
            // Get a free license here:
            // https://www.syncfusion.com/products/communitylicense
            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
                .AddJsonFile("appsettings.json", optional: false,
                reloadOnChange: true);
            var Configuration = builder.Build();
            var SyncfusionLicense = Configuration.GetSection("SyncfusionLicense");
            if (SyncfusionLicense != null)
            {
                Syncfusion.Licensing.SyncfusionLicenseProvider
                    .RegisterLicense(SyncfusionLicense.Value);
            }
        }

        public void ConfigureMvc(IMvcBuilder mvcBuilder)
        {
        }
    }
}
```

Also see: [Using Syncfusion In Oqtane Modules](https://blazorhelpwebsite.com/ViewBlogPost/47)

**Reply by Gotiap** [reply]
Thank you for your help and time @ADefWebserver ,

But I want to get access those same settings into ConfigureServices, As I want to register httpclient and need to use prod, stage or dev environment with different different urls.

Can you please try to use into ConfigureServices?

I read that blogs from [https://blazorhelpwebsite.com/](https://blazorhelpwebsite.com/)

Thank you.

**Reply by ADefWebserver** [reply]
> Can you please try to use into ConfigureServices?

I could not get that to work.

**Reply by sbwalker** [reply]
@Gotiap Have you looked at the Startup.cs class in the core framework? In your ConfigureServices method why could you not do something like:

```
                services.AddScoped(s =>
                {
                    var config = s.GetRequiredService<IConfigurationRoot>();
                    var client = new HttpClient(new HttpClientHandler { UseCookies = false });
                    client.BaseAddress = new Uri(config.GetSection("ServiceUrl").Value);
                    return client;
                });
```

**Reply by qmmughal** [reply]
Thank you guys for all options and suggestions
